Improve Configurability & Additional Fixes

tomalok opened the pull request:
Combining PR's #7, #8, #9, and #10.

tomalok commented on make_ami.sh:
```diff
@@ -1,9 +1,9 @@
 #!/bin/sh
-# vim:set ts=4:
+# vim: set ts=4 noet:
```
There were quite a few hard tabs still in the script, so rolled with it in absence of "et" in the vim options line.  I'll get that cleaned up.

mcrute replied:
Ah, my mistake then. I normally have `et` set in my default vim config but I started authoring this file on a machine that was not my normal one. Thanks for fixing it.

tomalok commented on make_ami.sh:
```diff
@@ -62,24 +62,24 @@ fetch_apk_tools() {
 	cat > "$target"/etc/apk/repositories <<-EOF
 	http://dl-cdn.alpinelinux.org/alpine/v$ALPINE_RELEASE/main
 	http://dl-cdn.alpinelinux.org/alpine/v$ALPINE_RELEASE/community
-	@testing http://dl-cdn.alpinelinux.org/alpine/edge/testing
```
good call.  will fix.

tomalok commented on make_ami.sh:
```diff
@@ -99,45 +99,44 @@ setup_chroot() {
-		acct \
+		linux-"$flavor" \
 		alpine-mirrors \
-		aws-ena-driver-vanilla@testing \
```
I prefer to use the smaller 'virt' kernel (and smaller instances), unfortunately there's no matching aws-ena-driver package for that (though I submitted https://github.com/alpinelinux/aports/pull/5023 to add a 'virt' flavor to the build), so (IIRC) adding this package also pulled the whole linux-vanilla, too.

mcrute replied:
I'm happy to merge this pending re-inclusion once alpinelinux/aports#5023 is merged. Although until then the virt kernel won't be able to use any of the enhanced networking support in newer series instances, and if I remember correctly, networking will not work at all without this driver on the 5-series instances.

Longer term it's probably worth flipping this whole image to use the virt kernel given the runtime environment.

tomalok commented on variables.json-default:
```diff
@@ -0,0 +1,22 @@
+  "ami_desc_suffix": " Release with EC2 Optimizations",
+  "kernel_flavor": "vanilla@edge-main",
+  "add_repos": "@edge-main http://dl-cdn.alpinelinux.org/alpine/edge/main,@edge-testing http://dl-cdn.alpinelinux.org/alpine/edge/testing",
+  "add_pkgs": "acct aws-ena-driver-vanilla@edge-testing e2fsprogs-extra",
```
The `e2fsprogs-extra` part was added without realizing that the alpine package was 1.0.0, while the latest release on the `tiny-ec2-bootstrap` repo was 1.1.0.  certainly the way to fix it is via the declared dependencies in alpinelinux/aports -- which doesn't appear to be in https://github.com/alpinelinux/aports/pull/5041 yet.  I had it in PR https://github.com/alpinelinux/aports/pull/5034 which I closed in favor of the bump to 1.2.0, and https://github.com/alpinelinux/aports/pull/5025 was closed in favor of 5034...

slightly OT -- how long does it generally take the alpinelinux folks to review/apply PRs, and what's the process for getting stuff out of edge/testing and into main or community?

mcrute replied:
It looks like I forgot to add `e2fsprogs-extra` in alpinelinux/aports#5041 but I've updated it now.

Normally they merge pretty quickly but things seem to be taking a while lately. The maintainers must be busy. I'll see if I can convince anyone to merge them on IRC this weekend when I have a little more time.

To get something moved from testing to edge generally just requires the author to do enough testing to be able to vouch for the package inclusion. I've not had a lot of time to do that testing lately which is why this is lagging. Inclusion from edge to main generally happens when new releases are cut but can be done out of cycle if there's a compelling reason to do it, but that has to be coordinated with the Alpine maintainers.

tomalok commented on variables.json-default:
```diff
@@ -0,0 +1,22 @@
+  "security_group": "",
+  "public_ip": "false",
+  "build_instance_type": "t2.nano",
+  "volume_name": "/dev/xvdf"
```
Yep, these probably don't need to be configurable -- just transferred from the Packer variables section, IIRC.

tomalok commented:
Limited access to the intarwebz this week, I'll work on addressing the requested changes next week-ish.

mcrute commented:
I'll pull in your changes and start to do some testing as well this weekend. I've been meaning to spend more time on getting this image production ready but work has been quite busy the past few months. I'd like to get the ena driver in main and the ec2-bootstrap patches merged then update my testing environment and let the image bake for a week or so before pushing a more broad release. Any testing you're willing to do would be helpful as well. Thanks!

tomalok commented on variables.json-default:
```diff
@@ -0,0 +1,22 @@
+  "ami_name_prefix": "Alpine-",
+  "ami_name_suffix": "-EC2",
+  "ami_desc_prefix": "Alpine Linux ",
+  "ami_desc_suffix": " Release with EC2 Optimizations",
```
FWIW, I'm using them...  :)

I added prefix/suffix vars so that when folks are building custom AMIs with this, that they have an easy way to avoid clashing with officially-built AMIs.  Wanted to avoid putting configurable things in multiple files.

[just noticed i hadn't clicked the comment button yesterday...]

tomalok commented:
Got most/all things you mentioned addressed, I think...  Will be taking a closer look as I do a little testing and rephrasing of comments before submitting an update.  A couple other bits...

* fixed NTP source (`chrony.conf` was using `pool`, not `server`)
* added an option for encrypting the AMI.
* discovered that SR-IOV drivers are provided with the 'virt' kernel, so we'll make sure that's noted in the AMI as well, regardless of kernel flavor.
* Considering adding something along the lines of `add_svc` for when someone adds a package and also wants it to start automatically.  Will probably leave it out for this round.

The T3 instance family appears to have landed recently, too.  I've just spun one up to take a look around -- it also requires ENA and NVMe.  I almost exclusively use smaller instances, so this bumps up my priority of getting NVMe into the virt flavor of the kernel.  Might have to dust off an IRC client?

t3.nano, linux-vanilla@edge-main
```
alpine@ip-192-168-3-137:~$ df -h
Filesystem                Size      Used Available Use% Mounted on
devtmpfs                 10.0M         0     10.0M   0% /dev
shm                     234.0M         0    234.0M   0% /dev/shm
/dev/nvme0n1            975.9M    531.7M    377.0M  59% /
tmpfs                    46.8M     92.0K     46.7M   0% /run
df: /sys/kernel/debug/tracing: Permission denied
alpine@ip-192-168-3-137:~$ free
             total       used       free     shared    buffers     cached
Mem:        479308      50508     428800         92       1800       8988
-/+ buffers/cache:      39720     439588
Swap:            0          0          0
alpine@ip-192-168-3-137:~$ uname -a
Linux ip-192-168-3-137 4.14.67-0-vanilla #1-Alpine SMP Mon Aug 27 07:53:04 UTC 2018 x86_64 Linux
```

t2.nano, linux-virt
```
alpine@ip-192-168-3-46:~$ df -h
Filesystem                Size      Used Available Use% Mounted on
devtmpfs                 10.0M         0     10.0M   0% /dev
shm                     243.6M         0    243.6M   0% /dev/shm
/dev/xvda               975.9M     85.1M    823.6M   9% /
tmpfs                    48.7M     88.0K     48.6M   0% /run
df: /sys/kernel/debug/tracing: Permission denied
alpine@ip-192-168-3-46:~$ free
             total       used       free     shared    buffers     cached
Mem:        498980      48860     450120         88       2180       8560
-/+ buffers/cache:      38120     460860
Swap:            0          0          0
alpine@ip-192-168-3-46:~$ uname -a
Linux ip-192-168-3-46 4.14.52-0-virt #1-Alpine SMP Tue Jun 26 07:21:56 UTC 2018 x86_64 Linux
```

(not entirely sure why `busybox`'s `df` wants to access /sys/kernel/debug/tracing.)

mcrute commented:
Thanks, this all looks good so I'm going to merge it. I didn't get a chance to test last weekend but I'll try again later this week. I use a fair number of t-series instances as well so it'll be high priority to get ENA working on that.

I'm also considering converting the `variables.json` file to YAML similar to the way the packer file is converted. It's just nicer to have a file format that allows comments. I'll probably add something like CloudFormation `!Join` as well to allow the `deploy_regions` list to be an actual list. Will tag you on the PR once I get it all added.
